### PR TITLE
Refactoring start

### PR DESCRIPTION
**Pull request by a-ok123**

### CODE REVIEW

File: utils/utils.go [hunk 1]
```diff
@@ -447,3 +450,40 @@ func GetServiceConfig(name string, format string, value interface{}) (string, er
 
 	return buf.String(), nil
 }
+
+// ReadStrings reads string???
+func ReadStrings(comment string) (string, error) {
```
Comment (pjenys0) [inline utils/utils.go:455]: I think should change to ReadUserInput()?

Comment (pjenys0) [reply utils/utils.go:455]: And this function just read a line, not multiple lines, so ReadString**s** is confusing

File: cmd/start.go [hunk 2]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
+			"extCfg":     "",
+			"extKey":     pastelid,
+		},
 	}
-	var aliasStatus map[string]interface{}
 
-	if err = json.Unmarshal([]byte(output), &aliasStatus); err != nil {
-		return err
+	if isCreate {
+		data, _ := json.Marshal(confData)
```
Comment (pjenys0) [inline cmd/start.go:716]: nit: should check error. but anw, can discard it now.

File: cmd/start.go [hunk 3]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
-	log.WithContext(ctx).Info("Remote:::Waiting for supernode started...")
-	time.Sleep(5000 * time.Millisecond)
+	json.Unmarshal([]byte(confFile), &conf)
 
-	log.WithContext(ctx).Info("Remote:::Supernode was started successfully")
-	return nil
-}
+	for k := range confData {
+		if conf[k] != nil {
+			confDataValue := confData[k].(map[string]string)
```
Comment (pjenys0) [inline cmd/start.go:788]: it's better should ensure can casts successful.
like :
```
confDataValue, ok := confData[k].(map[string]string)
if !ok {
  //handle error
}
```

File: cmd/start.go [hunk 4]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
+	} else {
+		masternodeConfPath = filepath.Join(workDirPath, "masternode.conf")
+	}
 
-	remoteWorkDirPath, remotePastelExecPath, remoteOsType, err := getRemoteInfo(config, client)
+	// Read ConfData from masternode.conf
+	confFile, err := ioutil.ReadFile(masternodeConfPath)
 	if err != nil {
-		return err
+		return false, err
```
Comment (pjenys0) [inline cmd/start.go:779]: it's better to wrap error with detail off error.
`return false, errros.Errorf("failed to read masternode cofinig file: %v", err)`

File: cmd/start.go [hunk 5]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
-	var pastelDPath string
+				}
+			}
+		}
+	}
 
-	if _, pastelDPath, _, _, err = checkPastelInstallPath(ctx, config, ""); err != nil {
-		return errNotFoundPastelPath
+	var updatedConf []byte
+	if updatedConf, err = json.Marshal(conf); err != nil {
```
Comment (pjenys0) [inline cmd/start.go:800]: it's better to wrap error with detail off error.
return false, errros.Errorf("failed to marshal master config: %v", err)

File: cmd/start.go [hunk 6]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
+		return false, err
 	}
 
-	if err = checkStartNodeParams(ctx, config); err != nil {
-		return err
+	if ioutil.WriteFile(masternodeConfPath, updatedConf, 0644) != nil {
+		return false, err
```
Comment (pjenys0) [inline cmd/start.go:805]: it's better to wrap error with detail off error.
return false, errros.Errorf("failed to write masternode cofinig file: %v", err)

File: cmd/start.go [hunk 7]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
-			go RunPasteld(ctx, config, fmt.Sprintf("--externalip=%s", flagNodeExtIP), "--daemon")
+		masternodeConfPath = filepath.Join(workDirPath, "masternode.conf")
+		masternodeConfPathBackup = filepath.Join(workDirPath, "masternode_%s.conf")
+	}
+	if _, err := os.Stat(masternodeConfPath); err == nil { // if masternode.conf File exists , backup
+		oldFileName := masternodeConfPath
+		currentTime := time.Now()
+		backupFileName := fmt.Sprintf(masternodeConfPathBackup, currentTime.Format("2021-01-01-23-59-59"))
+		if err := os.Rename(oldFileName, backupFileName); err != nil {
+			return err
```
Comment (pjenys0) [inline cmd/start.go:827]: nit:
oldFileName -> oldFilePath
backupFileName -> backupFilePath

**Review by pjenys0 — APPROVED** [review]
There are some minor findings. overall it looks good.

File: configs/config.go [hunk 8]
```diff
@@ -43,8 +43,8 @@ pastel-api:
   username: "{{.PastelUserName}}"
   password: "{{.PastelPassword}}"
 node:
-  pastel_id: jXXzhJJnfEk4cfJWeyX3t6o6EGBaCKsp6Qn38BUX62gHbBB325t1JqrkisKJGLopZ6QjzfeAmE623oK7GqS3sw 
-  pass_phrase: "passphrase"
+  pastel_id: {{.PasteID}}
```
Comment (georgkcode0) [inline configs/config.go:46]: carefully, `PasteID` and `Passphrase` are `string`

File: cmd/start.go [hunk 9]
```diff
@@ -218,964 +196,924 @@ func setupStartCommand() *cli.Command {
...
-	if err != nil {
-		return errors.Errorf("failed to convert %s to integer: %v", pastelID, err)
+func runPastelWalletNode(ctx context.Context, config *configs.Config) error {
+
+	walletnodeExecName := constants.WalletNodeExecName[utils.GetOS()]
+	log.WithContext(ctx).Infof("Starting walletnode - %s", walletnodeExecName)
+
+	var wnServiceArgs []string
+	wnServiceArgs = append(wnServiceArgs,
+		fmt.Sprintf("--config-file=%s", filepath.Join(config.WorkingDir, "wallet.yml")))
```
Comment (georgkcode0) [inline cmd/start.go:343]: incorrect file name, please change to `walletnode.yml`

File: cmd/start.go [hunk 10]
```diff
@@ -138,15 +115,16 @@ func setupStartSubCommand(config *configs.Config,
...
+		cli.NewFlag("remote", &flagMasterNodeColdHot),
 		cli.NewFlag("ssh-ip", &flagMasterNodeSSHIP).
 			SetUsage(green("remote supernode specific: Required, SSH address of the remote HOT node")),
 		cli.NewFlag("ssh-port", &flagMasterNodeSSHPort).
 			SetUsage(green("remote supernode specific: Optional, SSH port of the remote HOT node")).SetValue(22),
-		cli.NewFlag("remote-work-dir", &config.RemoteWorkingDir).
-			SetUsage(green("remote supernode specific: Required, location of the working directory")),
-		cli.NewFlag("pastelpath", &flagMasterNodePastelPath).
-			SetUsage(green("remote supernode specific: Optional, The path of the pasteld")),
+		cli.NewFlag("remote-dir", &config.RemotePastelExecDir).SetAliases("d").
```
Comment (georgkcode0) [inline cmd/start.go:124]: duplicate flag `d`

File: cmd/start.go [hunk 11]
```diff
@@ -138,15 +115,16 @@ func setupStartSubCommand(config *configs.Config,
...
+			SetUsage(green("Optional, Location where of pastel node directory on the remote computer (default: $HOME/pastel-utility)")),
+		cli.NewFlag("work-dir", &config.RemoteWorkingDir).SetAliases("w").
```
Comment (georgkcode0) [inline cmd/start.go:126]: duplicate flag `work-dir`

Comment (georgkcode0) [reply cmd/start.go:126]: duplicate flag `w`

File: utils/utils.go (hunk 1 above)
Comment (a-ok123) [reply utils/utils.go:455]: tbh i don't even understand the purpose of this function

**Review by pjenys0 — APPROVED** [review]

**Pull request merged**